common/src/main/java/io/netty/util/internal/logging/MessageFormatter.java

```diff
@@ -307,120 +284,139 @@ private static void safeObjectAppend(StringBuffer sbuf, Object o) {
         }
     }
 
-    private static void objectArrayAppend(StringBuffer sbuf, Object[] a,
-                                          Map<Object[], Void> seenMap) {
-        sbuf.append('[');
-        if (!seenMap.containsKey(a)) {
-            seenMap.put(a, null);
-            final int len = a.length;
-            for (int i = 0; i < len; i++) {
-                deeplyAppendParameter(sbuf, a[i], seenMap);
-                if (i != len - 1) {
-                    sbuf.append(", ");
+    private static void objectArrayAppend(StringBuilder sbuf, Object[] a, Set<Object[]> seenSet) {
+        int iMax = a.length - 1;
+        if (iMax == -1) {
+            return;
+        }
+        if (seenSet == null) {
+            seenSet = new HashSet<Object[]>(a.length);
+        }
+        if (seenSet.add(a)) {
+            for (int i = 0; ; i++) {
+                deeplyAppendParameter(sbuf, a[i], seenSet);
+                if (i == iMax) {
+                    break;
                 }
+                sbuf.append(", ");
             }
             // allow repeats in siblings
-            seenMap.remove(a);
+            seenSet.remove(a);
         } else {
             sbuf.append("...");
         }
-        sbuf.append(']');
     }
 
-    private static void booleanArrayAppend(StringBuffer sbuf, boolean[] a) {
-        sbuf.append('[');
-        final int len = a.length;
-        for (int i = 0; i < len; i++) {
+    private static void booleanArrayAppend(StringBuilder sbuf, boolean[] a) {
+        int iMax = a.length - 1;
+        if (iMax == -1) {
+            return;
+        }
+        for (int i = 0; ; i++) {
             sbuf.append(a[i]);
-            if (i != len - 1) {
-                sbuf.append(", ");
+            if (i == iMax) {
+                return;
             }
+            sbuf.append(", ");
         }
-        sbuf.append(']');
     }
 
-    private static void byteArrayAppend(StringBuffer sbuf, byte[] a) {
-        sbuf.append('[');
-        final int len = a.length;
-        for (int i = 0; i < len; i++) {
+    private static void byteArrayAppend(StringBuilder sbuf, byte[] a) {
+        int iMax = a.length - 1;
+        if (iMax == -1) {
+            return;
+        }
+        for (int i = 0; ; i++) {
             sbuf.append(a[i]);
-            if (i != len - 1) {
-                sbuf.append(", ");
+            if (i == iMax) {
+                return;
             }
+            sbuf.append(", ");
         }
-        sbuf.append(']');
     }
 
-    private static void charArrayAppend(StringBuffer sbuf, char[] a) {
-        sbuf.append('[');
-        final int len = a.length;
-        for (int i = 0; i < len; i++) {
+    private static void charArrayAppend(StringBuilder sbuf, char[] a) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)